File: src/cli/docker.go
package cli

import (
	"fmt"
	"strings"
)

// Docker wraps Docker commands
type Docker struct {
	dockerCmd   string
	showAll     bool
	includeAll  bool
	force       bool
	tail        int
}

// NewDocker creates a new Docker command builder
func NewDocker() *Docker {
	return &Docker{
		dockerCmd: "docker",
		tail:      20,
	}
}

// DockerCmd sets docker command path
func (d *Docker) DockerCmd(cmd string) *Docker {
	d.dockerCmd = cmd
	return d
}

// ShowAll shows all containers
func (d *Docker) ShowAll(value bool) *Docker {
	d.showAll = value
	return d
}

// IncludeAll includes all in prune
func (d *Docker) IncludeAll(value bool) *Docker {
	d.includeAll = value
	return d
}

// Force sets force flag
func (d *Docker) Force(value bool) *Docker {
	d.force = value
	return d
}

// Tail sets tail lines
func (d *Docker) Tail(lines int) *Docker {
	d.tail = lines
	return d
}

// FindDocker generates command to find docker command path
func (d *Docker) FindDocker() string {
	return "dpkg -L docker.io 2>/dev/null | grep -E '/bin/docker$' | head -1 || dpkg -L docker-ce 2>/dev/null | grep -E '/bin/docker$' | head -1 || command -v docker 2>/dev/null || which docker 2>/dev/null || find /usr /usr/local -name docker -type f 2>/dev/null | head -1 || test -x /usr/bin/docker && echo /usr/bin/docker || test -x /usr/local/bin/docker && echo /usr/local/bin/docker || echo 'docker'"
}

// Version generates command to get Docker version
func (d *Docker) Version() string {
	return fmt.Sprintf("%s --version 2>&1", d.dockerCmd)
}

// PS generates command to list containers
func (d *Docker) PS() string {
	allFlag := ""
	if d.showAll {
		allFlag = "-a"
	}
	return fmt.Sprintf("%s ps %s 2>&1", d.dockerCmd, allFlag)
}

// SwarmInit generates command to initialize Docker Swarm
func (d *Docker) SwarmInit(advertiseAddr string) string {
	return fmt.Sprintf("%s swarm init --advertise-addr %s 2>&1", d.dockerCmd, advertiseAddr)
}

// SwarmJoinToken generates command to get Swarm join token
func (d *Docker) SwarmJoinToken(role string) string {
	return fmt.Sprintf("%s swarm join-token %s -q 2>&1", d.dockerCmd, role)
}

// SwarmJoin generates command to join Docker Swarm
func (d *Docker) SwarmJoin(token, managerAddr string) string {
	return fmt.Sprintf("%s swarm join --token %s %s 2>&1", d.dockerCmd, token, managerAddr)
}

// NodeLS generates command to list Swarm nodes
func (d *Docker) NodeLS() string {
	return fmt.Sprintf("%s node ls 2>&1", d.dockerCmd)
}

// NodeUpdate generates command to update node availability
func (d *Docker) NodeUpdate(nodeName, availability string) string {
	return fmt.Sprintf("%s node update --availability %s %s 2>&1", d.dockerCmd, availability, nodeName)
}

// VolumeCreate generates command to create Docker volume
func (d *Docker) VolumeCreate(volumeName string) string {
	return fmt.Sprintf("%s volume create %s 2>/dev/null || true", d.dockerCmd, volumeName)
}

// VolumeRM generates command to remove Docker volume
func (d *Docker) VolumeRM(volumeName string) string {
	forceFlag := ""
	if d.force {
		forceFlag = "-f "
	}
	return fmt.Sprintf("%s volume rm %s%s 2>/dev/null || true", d.dockerCmd, forceFlag, volumeName)
}

// Run generates command to run Docker container
func (d *Docker) Run(image, name string, args map[string]interface{}) string {
	cmd := fmt.Sprintf("%s run -d --name %s", d.dockerCmd, name)
	if restart, ok := args["restart"].(string); ok {
		cmd += fmt.Sprintf(" --restart=%s", restart)
	}
	if network, ok := args["network"].(string); ok {
		cmd += fmt.Sprintf(" --network %s", network)
	}
	if volumes, ok := args["volumes"].([]string); ok {
		for _, vol := range volumes {
			cmd += fmt.Sprintf(" -v %s", vol)
		}
	}
	if ports, ok := args["ports"].([]string); ok {
		for _, port := range ports {
			cmd += fmt.Sprintf(" -p %s", port)
		}
	}
	if securityOpts, ok := args["security_opts"].([]string); ok {
		for _, opt := range securityOpts {
			cmd += fmt.Sprintf(" --security-opt %s", opt)
		}
	}
	cmd += fmt.Sprintf(" %s", image)
	if commandArgs, ok := args["command_args"].([]string); ok {
		for _, arg := range commandArgs {
			cmd += fmt.Sprintf(" %s", quote(arg))
		}
	}
	cmd += " 2>&1"
	return cmd
}

// Stop generates command to stop Docker container
func (d *Docker) Stop(containerName string) string {
	return fmt.Sprintf("%s stop %s 2>/dev/null || true", d.dockerCmd, containerName)
}

// RM generates command to remove Docker container
func (d *Docker) RM(containerName string) string {
	return fmt.Sprintf("%s rm %s 2>/dev/null || true", d.dockerCmd, containerName)
}

// Logs generates command to get Docker container logs
func (d *Docker) Logs(containerName string) string {
	return fmt.Sprintf("%s logs %s 2>&1 | tail -%d", d.dockerCmd, containerName, d.tail)
}

// SystemPrune generates command to prune Docker system
func (d *Docker) SystemPrune() string {
	flags := ""
	if d.includeAll {
		flags += " -a"
	}
	if d.force {
		flags += " -f"
	}
	return fmt.Sprintf("%s system prune%s 2>/dev/null || true", d.dockerCmd, flags)
}

// IsInstalledCheck generates command to check if Docker is installed
func (d *Docker) IsInstalledCheck() string {
	return fmt.Sprintf("command -v %s >/dev/null 2>&1 && echo installed || echo not_installed", d.dockerCmd)
}

// ParseDockerIsInstalled parses output to check if Docker is installed
func ParseDockerIsInstalled(output string) bool {
	return strings.Contains(strings.ToLower(output), "installed")
}

File: csharp-enva/CLI/Docker.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Enva.CLI;

public class Docker
{
    private string dockerCmd = "docker";
    private bool showAll;
    private bool includeAll;
    private bool force;
    private int tail = 20;

    public static Docker NewDocker()
    {
        return new Docker();
    }

    public Docker DockerCmd(string cmd)
    {
        dockerCmd = cmd;
        return this;
    }

    public Docker ShowAll(bool value)
    {
        showAll = value;
        return this;
    }

    public Docker IncludeAll(bool value)
    {
        includeAll = value;
        return this;
    }

    public Docker Force(bool value)
    {
        force = value;
        return this;
    }

    public Docker Tail(int lines)
    {
        tail = lines;
        return this;
    }

    public string FindDocker()
    {
        return "dpkg -L docker.io | grep -E '/bin/docker$' | head -1 || dpkg -L docker-ce | grep -E '/bin/docker$' | head -1 || command -v docker || which docker || find /usr /usr/local -name docker -type f | head -1 || test -x /usr/bin/docker && echo /usr/bin/docker || test -x /usr/local/bin/docker && echo /usr/local/bin/docker || echo 'docker'";
    }

    public string Version()
    {
        return $"{dockerCmd} --version";
    }

    public string PS()
    {
        string allFlag = showAll ? "-a" : "";
        return $"{dockerCmd} ps {allFlag}".Trim();
    }

    public string SwarmInit(string advertiseAddr)
    {
        return $"{dockerCmd} swarm init --advertise-addr {advertiseAddr}";
    }

    public string SwarmJoinToken(string role)
    {
        return $"{dockerCmd} swarm join-token {role} -q";
    }

    public string SwarmJoin(string token, string managerAddr)
    {
        return $"{dockerCmd} swarm join --token {token} {managerAddr}";
    }

    public string NodeLS()
    {
        return $"{dockerCmd} node ls";
    }

    public string NodeUpdate(string nodeName, string availability)
    {
        return $"{dockerCmd} node update --availability {availability} {nodeName}";
    }

    public string VolumeCreate(string volumeName)
    {
        return $"{dockerCmd} volume create {volumeName} || true";
    }

    public string VolumeRM(string volumeName)
    {
        string forceFlag = force ? "-f " : "";
        return $"{dockerCmd} volume rm {forceFlag}{volumeName} || true";
    }

    public string Run(string image, string name, Dictionary<string, object> args)
    {
        string cmd = $"{dockerCmd} run -d --name {name}";
        if (args != null)
        {
            if (args.ContainsKey("restart") && args["restart"] is string restart)
            {
                cmd += $" --restart={restart}";
            }
            if (args.ContainsKey("network") && args["network"] is string network)
            {
                cmd += $" --network {network}";
            }
            if (args.ContainsKey("volumes") && args["volumes"] is List<string> volumes)
            {
                foreach (var vol in volumes)
                {
                    cmd += $" -v {vol}";
                }
            }
            if (args.ContainsKey("ports") && args["ports"] is List<string> ports)
            {
                foreach (var port in ports)
                {
                    cmd += $" -p {port}";
                }
            }
            if (args.ContainsKey("security_opts") && args["security_opts"] is List<string> securityOpts)
            {
                foreach (var opt in securityOpts)
                {
                    cmd += $" --security-opt {opt}";
                }
            }
        }
        cmd += $" {image}";
        if (args != null && args.ContainsKey("command_args") && args["command_args"] is List<string> commandArgs)
        {
            foreach (var arg in commandArgs)
            {
                cmd += $" {Quote(arg)}";
            }
        }
        return cmd;
    }

    public string Stop(string containerName)
    {
        return $"{dockerCmd} stop {containerName} || true";
    }

    public string RM(string containerName)
    {
        return $"{dockerCmd} rm {containerName} || true";
    }

    public string Logs(string containerName)
    {
        return $"{dockerCmd} logs {containerName} | tail -{tail}";
    }

    public string SystemPrune()
    {
        string flags = "";
        if (includeAll)
        {
            flags += " -a";
        }
        if (force)
        {
            flags += " -f";
        }
        return $"{dockerCmd} system prune{flags} || true";
    }

    public string IsInstalledCheck()
    {
        return $"command -v {dockerCmd} && echo installed || echo not_installed";
    }

    public static bool ParseDockerIsInstalled(string output)
    {
        return output.ToLower().Contains("installed");
    }

    private string Quote(string s)
    {
        if (s.Contains(" ") || s.Contains("$") || s.Contains("'"))
        {
            return $"'{s.Replace("'", "'\"'\"'")}'";
        }
        return s;
    }
}
